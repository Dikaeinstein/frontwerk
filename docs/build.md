# frontwerk üõ†Ô∏è builds your app

Build your app with one of the following:

* [BabelJS][babel]
* [RollupJS][rollup]
* [Webpack][webpack]

## Choose your builder

**Babel (default)**

```json
{
  "scripts": {
    "build": "frontwerk build"
  }
}
```

Continue reading about [using babel][docs-babel].

**Rollup**

```json
{
  "scripts": {
    "build": "frontwerk build --bundle"
  }
}
```

Continue reading about [using rollup][docs-rollup].

**Webpack**

```json
{
  "scripts": {
    "build": "frontwerk build --pack"
  }
}
```

Continue reading about [using webpack][docs-webpack].

[babel]: https://babeljs.io/
[rollup]: https://rollupjs.org/
[webpack]: https://webpack.js.org/
[docs-babel]: https://github.com/tricinel/frontwerk/blob/master/docs/babel.md
[docs-rollup]: https://github.com/tricinel/frontwerk/blob/master/docs/rollup.md
[docs-webpack]: https://github.com/tricinel/frontwerk/blob/master/docs/webpack.md
