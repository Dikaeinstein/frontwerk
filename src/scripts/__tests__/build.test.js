/* eslint-disable global-require */
import {
  unquoteSerializer,
  winPathSerializer
} from '../../helpers/serializers';

jest.mock('cross-spawn');

const cases = require('jest-in-case');
const jestSerializerPath = require('jest-serializer-path');

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

const testCases = [
  {
    name: 'calls node with default args'
  },
  {
    name: 'calls node with --bundle',
    args: ['--bundle']
  },
  {
    name: 'calls node with --pack',
    args: ['--pack']
  }
];

const testFn = ({ args = [] }) => {
  const { sync: crossSpawnSyncMock } = require('cross-spawn');
  const originalExit = process.exit;
  const originalArgv = process.argv;
  process.exit = jest.fn();

  try {
    // tests
    process.argv = ['node', '../build', ...args];
    crossSpawnSyncMock.mockClear();
    require('../build');

    expect(crossSpawnSyncMock).toHaveBeenCalledTimes(1);
    const [firstCall] = crossSpawnSyncMock.mock.calls;
    const [script, calledArgs] = firstCall;
    expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
  } catch (error) {
    throw error;
  } finally {
    // afterEach
    process.exit = originalExit;
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('build', testFn, testCases);
