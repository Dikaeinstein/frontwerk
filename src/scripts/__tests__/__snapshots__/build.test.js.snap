// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format calls node with --bundle 1`] = `"concurrently --names esm,cjs,umd,umd.min \\"cross-env BUILD_ROLLUP=true BUILD_FORMAT=esm BUILD_MINIFY=false NODE_ENV=development  rollup --config ./src/config/rollup.config.js\\" \\"cross-env BUILD_ROLLUP=true BUILD_FORMAT=cjs BUILD_MINIFY=false NODE_ENV=development  rollup --config ./src/config/rollup.config.js\\" \\"cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=false NODE_ENV=development  rollup --config ./src/config/rollup.config.js --sourcemap\\" \\"cross-env BUILD_ROLLUP=true BUILD_FORMAT=umd BUILD_MINIFY=true NODE_ENV=production  rollup --config ./src/config/rollup.config.js --sourcemap\\""`;

exports[`format calls node with --pack 1`] = `"concurrently --names webpack \\"cross-env BUILD_WEBPACK=true webpack --config ./src/config/webpack.config.js\\""`;

exports[`format calls node with default args 1`] = `"babel --out-dir dist --copy-files --ignore __tests__,__mocks__ --presets /Users/tricinel/Work/Personal/repos/ff-scripts/packages/frontwerk/src/config/babelrc.js src"`;
